src/service/history-impl.js

```diff
@@ -601,14 +611,25 @@ export class HistoryBindingVirtual_ {
 
   /** @override */
   push() {
-    // Current implementation doesn't wait for response from viewer.
-    this.updateStackIndex_(this.stackIndex_ + 1);
-    this.viewer_.postPushHistory(this.stackIndex_);
-    return Promise.resolve(this.stackIndex_);
+    // Waiting for popping confirmation to happen otherwise a race condition
+    // between popping and pushing might happen causing history to be messed
+    // up. This will only wait 100ms to avoid getting stuck if confirmation
+    // never arrived.
+    return Promise.race([this.timer_.promise(100), this.awaitPoppingPromise_])
```

Comment:
Why `awaitPoppingPromise_` necessary? Isn't queuing feature already provided by the parent `History` class? It'd seem all we need to do is return a promise from viewer's `postPushHistory` call, which may already be the case.